src/eliminate_contiguous.cpp

```diff
@@ -70,53 +70,79 @@ static bool try_compute_shape(instruction_ref ins,
     return try_compute_shape(ins, inputs, mods);
 }
 
-void eliminate_contiguous::apply(module& m) const
+template <class F>
+static void remove_contiguous(const std::string& op_name, module& m, F f)
 {
-    std::vector<instruction_ref> const_instruction;
+    auto last = std::prev(m.end());
+    std::vector<instruction_ref> const_instructions;
 
     for(auto ins : iterator_for(m))
     {
         // return instruction should have inputs with standard shape
         if(ins->name() == "@return")
             continue;
 
+        if(ins != last and ins->outputs().empty())
+            continue;
+
+        if(not f(ins))
+            continue;
+
         // Make a copy so we can modify it while we iterate
         auto args     = ins->inputs();
         auto new_args = args;
         auto mod_args = ins->module_inputs();
 
         for(auto arg : ins->inputs())
         {
-            if(arg->name() == op_name)
+            if(arg->name() != op_name)
+                continue;
+            auto prev = arg->inputs().front();
+            replace(new_args, arg, prev);
+            if(try_compute_shape(ins, new_args, mod_args))
+            {
+                instruction::replace_argument(ins, arg, prev);
+            }
+            else if(prev->can_eval())
             {
-                auto prev = arg->inputs().front();
                 replace(new_args, arg, prev);
                 if(try_compute_shape(ins, new_args, mod_args))
                 {
                     instruction::replace_argument(ins, arg, prev);
                 }
                 else if(prev->can_eval())
                 {
-                    const_instruction.push_back(arg);
+                    const_instructions.push_back(arg);
                 }
             }
         }
     }
 
     // Perform evaluations in parallel
-    std::vector<argument> literals(const_instruction.size());
-    par_for(const_instruction.size(), 1, [&](const auto i) {
+    std::vector<argument> literals(const_instructions.size());
```

Comment:
Why are we doing constant folding here and not in propagate constant pass?

Reply:
Propagate constant won't fold on a contiguous operator as it could be inefficient if contiguous can later be eliminated. So we const fold the contiguous here if it cant be removed.

Reply:
I find it better to rather augment propagate constant to handle this to keep logic separated.

Reply:
I agree, but I think we should have a seperate issue for that.